reporting/src/nav-tracking-detector.js

```diff
@@ -0,0 +1,261 @@
+    } else if (event.type === 'safe-search-landing') {
+      this._analyzeLanding(event.details);
+    }
+  }
+
+  // general case: page navigation
+  _analyzeNavigation(event) {
+    // TODO:
+    // A difference is that publically indexed 'search -> host' navigations
+    // less sensitive than arbitrary 'host -> host' navigations. Thus,
+    // it is likely that additional checks will be needed to support them.
+    // For now, start start without it.
+    console.debug('[STUB]: general navigation are not yet covered', event);
+  }
+
+  // special case: public search engine landings
+  _analyzeLanding({ from, to, redirects }) {
+    // Open questions:
+    // * Is it sufficient to look only at the first hop? For instance, what
+    //   about permanent a redirect that is still controlled by the original
+    //   site owner, before it hands control over to the tracker?
+    // * Currently, we look only at the first hop, even though there are chains
+    //   of tracker redirects? Potentially, these could be useful. The message
+    //   thus puts the results in an array; but for now, it will only have a
+    //   single entry.
+    // * Should we use the statusCode? For instance, treat permanent redirects
+    //   differently? - Currently, we do not.
+    const { category, query: unsafeQuery } = from;
+    const { isAd, trackingUrls } = isSearchAdRedirect(category, redirects);
+    if (!isAd) {
+      return;
+    }
+
+    // Since the context is a search engine landing, it is likely
+    // that it is a public hostname. Also, hostname are normally safe
+    // to share (e.g. they will be sent as cleartext even in https).
+    const hostname = tryParseHostname(to.targetUrl);
+    if (!hostname) {
+      return;
+    }
+    const trackingHosts = trackingUrls.map(tryParseHostname);
+
+    // null out the query if there is the risk of leaking information
+    const { accept } = this.sanitizer.checkSuspiciousQuery(unsafeQuery);
+    const query = accept ? unsafeQuery : null;
+
+    const action = 'wtm.nav-track-detect.search-ad';
+    this._registerJob({
+      type: 'nav-track-detect:quorum-isAdCheck',
+      args: {
+        action,
+        payload: {
+          from: {
+            search: {
+              category,
+              query,
+            },
+          },
+          to: {
+            hostname,
+          },
+          via: {
+            redirects: trackingHosts,
+          },
+        },
+        quorumCheck: JSON.stringify([
+          action,
+          category,
+          hostname,
+          trackingHosts,
+        ]),
+      },
+    });
+  }
+
+  async _passesQuorum(quorumCheck) {
+    requireString(quorumCheck);
+
+    // TODO: maybe break this also in two independent jobs
+    // (not strictly required, but could improve error recovery).
+    await this.quorumChecker.sendQuorumIncrement({ text: quorumCheck });
+    return this.quorumChecker.checkQuorumConsent({ text: quorumCheck });
```

Comment:
couldn't this be a single network request? Increment could respond with an information if quorum has been reached

Reply:
It is consistent with the existing use of the protocol. I do not know for certain if there was a reason for separating the operations. Arguably, the second operation could be delayed (e.g. to better detect new pages).